### ISSUE TITLE
Should reflection support routines for reasoning about routines accepting type arguments?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#21749]
While using `Reflection` today, I was initially surprised to find that I couldn't call:

```chapel
canResolve("max", myType)
```

where the reason is that we don't support `canResolve()` overloads that accept types.  Enabling my case looks to be trivial by adding the overload:

```chapel
proc canResolve(param fname: string, type t) param : bool
  return __primitive("call and fn resolves", fname, t);
```

which makes me inclined to propose we add such a routine.

The design challenge, though, is that it's much harder to support a canResolve() routine that takes multiple types or a mix of types and values, or types, values, and params (that maintains their paramness) for reasons similar to https://github.com/chapel-lang/chapel/issues/16917 and https://github.com/chapel-lang/chapel/issues/11800—namely, varargs routines don't accept types and params because tuples don't support types and params in a param-preserving way.

So, is it better to add the overload that I was looking for to improve the capabilities of the Reflection routine, or does it feel like too much of a one-off, and just starting us down a slippery slope of having a vague line between what is supported and what one might wish was supported?

**Comment by mppf** [comment]
I thought we were heading towards having a `canResolve(<expr>)` mechanism? See e.g. https://github.com/Cray/chapel-private/issues/2826 . Such a feature would also solve this problem -- you could just write `canResolve(max(myType))` (or whatever we call it).

**Comment by bradcray** [comment]
I'd forgotten about that, but agree that would be a silver bullet here.

**Comment by vasslitvinov** [comment]
We want both forms: one that checks if the entire expression is resolvable and the other to check a specific call. The first does not replace the second because it may produce a misleading result when a sub-expression is unintentionally broken.

For the second form, yes, we want it to accept all arguments, ex. const/non-const, param, type, and so on. Implementing that should be straightforward.

**Comment by vasslitvinov** [comment]
See also #11802 and #10918.

**Comment by jabraham17** [comment]
Commenting here as this would be a really nice thing to just support, and for params as well.

Currently, params give really confusing results. This program can be compiled and run.
```chapel
use Reflection;
proc bar(param i) {
  return i;
}
proc foo(param i) {
  if canResolve("bar", i) {
    writeln("bar cannot be resolved");
  } else {
    writeln("bar can be resolved");
  }
  writeln("calling bar ", bar(i));
}

foo(1);
```

The language allows passing `i` to `canResolve`, losing its param-ness. This then results in the very confusing behavior that `bar` can be resolved and called, but `canResolve` reports false.

**Comment by mppf** [comment]
Yes both the type and param challenges are stemming from problems in storing these in tuples. (Currently there are type tuples, but no mixes of values and types, and there are no param tuples). `canResolve` is implemented as a varargs function and varargs are (in principle) implemented as tuples. A related challenge is named arguments. Of course, we could change the way `canResolve` is implemented, or shore up tuples in this regard.

IMO it will be easier for us to get to "can this expression resolve?" being user facing. This mechanism is already used in the standard library, e.g.:

```
    iter these(param tag: iterKind) ref
      where tag == iterKind.standalone &&
            __primitive("resolves", _value.these(tag=tag)) { ... }
```

**Comment by vasslitvinov** [comment]
I like "can this expression resolve". However we need to retain`canResolve` for backward compatibility as well. There, I see that the current dropping param-ness behavior is a bug for us to fix.
